dask_deltatable/write.py

```diff
@@ -0,0 +1,262 @@
+            raise AssertionError("DeltaTable already exists.")
+        elif mode == "ignore":
+            return
+
+        current_version = table.version()
+
+        if partition_by:
+            assert partition_by == table.metadata().partition_columns
+        else:
+            partition_by = table.metadata().partition_columns
+
+        if table.protocol().min_writer_version > MAX_SUPPORTED_WRITER_VERSION:
+            raise DeltaProtocolError(
+                "This table's min_writer_version is "
+                f"{table.protocol().min_writer_version}, "
+                "but this method only supports version 2."
+            )
+    else:  # creating a new table
+        current_version = -1
+
+    # FIXME: schema is only known at this point if provided by the user
+    if partition_by and schema:
+        partition_schema = pa.schema([schema.field(name) for name in partition_by])
+        partitioning = ds.partitioning(partition_schema, flavor="hive")
+    else:
+        if partition_by:
+            raise NotImplementedError("Have to provide schema when using partition_by")
+        partitioning = None
+    if mode == "overwrite":
+        # FIXME: There are a couple of checks that are not migrated yet
+        raise NotImplementedError()
+
+    written = df.map_partitions(
+        _write_partition,
+        schema=schema,
+        partitioning=partitioning,
+        current_version=current_version,
+        file_options=file_options,
+        max_open_files=max_open_files,
+        max_rows_per_file=max_rows_per_file,
+        min_rows_per_group=min_rows_per_group,
+        max_rows_per_group=max_rows_per_group,
+        filesystem=filesystem,
+        max_partitions=max_partitions,
+        meta=(None, object),
+    )
+    final_name = "delta-commit"
+    dsk = {
+        (final_name, 0): (
+            _commit,
+            table,
+            written.__dask_keys__(),
+            table_uri,
+            schema,
+            mode,
+            partition_by,
+            name,
+            description,
+            configuration,
+            storage_options,
+            partition_filters,
+        )
+    }
+    graph = HighLevelGraph.from_collections(final_name, dsk, dependencies=(written,))
+    return Scalar(graph, final_name, "")
+
+
+def _commit(
+    table,
+    schemas_add_actions_nested,
+    table_uri,
+    schema,
+    mode,
+    partition_by,
+    name,
+    description,
+    configuration,
+    storage_options,
+    partition_filters,
+):
+    schemas = list(flatten(pluck(0, schemas_add_actions_nested)))
+    add_actions = list(flatten(pluck(1, schemas_add_actions_nested)))
+    # TODO: What should the behavior be if the schema is provided? Cast the
+    # data?
+    if schema:
+        schemas.append(schema)
+
+    # TODO: This is applying a potentially stricted schema control than what
+    # Delta requires but if this passes, it should be good to go
+    schema = validate_compatible(schemas)
```

Comment:
I took this code from https://github.com/data-engineering-collective/plateau where I've built a similar thing before. The challenge here is to get a schema from all the partitions that represents the transaction the best. this is particularly a struggle for partitions will null columns because pyarrow then does not give a proper schema. This method effectively looks at all written schemas, deduplicates it and merges them into a super-schema, i.e. it fills null columns with a proper type.

It also raises if there are incompatible schemas detected. Incompatible in this sense means, for example that the same column has an integer and a float in two different partitions. This _may_ be stricter than what delta requires and a lot of this is already covered by dask but I used this regardless since we need the scheme deduplication.

I'm very open to throwing this out again down the road but this gets us started quickly since this is battle-proven code.